Fixes for local builds

dberkholz opened the pull request:
Trying to get local builds working for people outside of Percona. Here's a couple of initial steps, which fix an initial build failure on pmm-client when a VictoriaMetrics submodule did not exist, and disable the S3 private cache by default (while creating a CACHE setting to enable it) — which was mandatory in build-server-rpm.

dberkholz committed:
Make builds work for anyone through disabling private cache by default.

The prior behavior of build-server-rpm used an AWS S3 cache only
accessible with a private key, with no way to disable it. This made it
impossible for anyone without that key to successfully run a build.

The new behavior disables this cache by default, so builds work for
everyone with no modification. It adds a 'cache' variable in the 'vars'
file, controlled by a CACHE environment variable, that allows setting
CACHE to anything besides the value 'false' to enable the private cache.

dberkholz committed:
Build fix: Add VictoriaMetrics submodule

On a clean checkout, pmm-client will not build without adding a
submodule for VictoriaMetrics.

JNKPercona commented:
server docker - perconalab/pmm-server-fb:PR-2406-f0731fb
client docker - perconalab/pmm-client-fb:PR-2406-f0731fb
client - https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2406-f0731fb.tar.gz
Create Staging Instance: https://pmm.cd.percona.com/job/aws-staging-start/parambuild/?DOCKER_VERSION=perconalab/pmm-server-fb:PR-2406-f0731fb&CLIENT_VERSION=https://s3.us-east-2.amazonaws.com/pmm-build-cache/PR-BUILDS/pmm2-client/pmm2-client-PR-2406-f0731fb.tar.gz

JNKPercona commented:
API tests have failed, Please check: API: https://pmm.cd.percona.com/job/pmm2-api-tests/9514/

JNKPercona commented:
UI tests have failed, Please check: UI: https://pmm.cd.percona.com/job/pmm2-ui-tests/13219/

BupycHuk commented on .gitmodules:
```diff
@@ -6,6 +6,10 @@
 
 # PMM Client
 
+[submodule "vmagent"]
+	path = sources/vmagent/src/github.com/VictoriaMetrics/VictoriaMetrics
+	url = https://github.com/VictoriaMetrics/VictoriaMetrics.git
+	branch = main
```
I don't think that this submodule is required

dberkholz replied:
Thanks for the review! Some sort of fix is necessary even if it's not this. I'll share some more detail on the issue and would appreciate your insights on the right solution.

Local build failed for me, on a clean clone, with the following error in a build log:

```
+ build_command='go build -o ${binary_dir}/bin/vmagent ./src/github.com/VictoriaMetrics/VictoriaMetrics/app/vmagent'
++ date +%s
++ cd sources/vmagent/src/github.com/VictoriaMetrics/VictoriaMetrics
./build/bin/build-client-binary: line 42: cd: sources/vmagent/src/github.com/VictoriaMetrics/VictoriaMetrics: No such file or directory
```

Line 42 is the docker run that's called from later in build-client-binary:
`gobuild_component "vmagent" "github.com/VictoriaMetrics/VictoriaMetrics" "app/vmagent"`

ademidoff replied:
@dberkholz Thanks for taking a moment and trying fix the local build.

Just wanted to let you know that our Q2 goal is to get rid of submodules in favor of a more flexible build helper tool which is `cy.yml`.

nikita-b replied:
We use a separate process for vmagent:
https://github.com/Percona-Lab/pmm-submodules/blob/890d99ab0e31867b5341f80feecf8789e6e7491b/build/bin/build-client-source#L45

We download the source code as an archive. I don't know why.

How do you run the building process?

dberkholz replied:
I run as per the docs [here](https://confluence.percona.com/pages/viewpage.action?spaceKey=PMM&title=Feature+Builds+-+Technical+design), `./build/bin/build`

nikita-b commented on build/bin/vars:
```diff
@@ -2,6 +2,9 @@ bin_dir=$(cd $(dirname $0); pwd -P)
 tmp_dir=${root_dir}/tmp
 
+# Use s3 cache for PMM server build (build-server-rpm)
+cache=${CACHE:-false}
```
I think it's better to have: `true` as the default value because we already have a couple of pipelines that rely on this behavior.

You can just use [Makefile](https://github.com/Percona-Lab/pmm-submodules/blob/PMM-2.0/Makefile) and set default values for manual building

dberkholz replied:
The most important thing here is that it works by default for non-Percona people. They won't have the patience to screw around with various config options on their very first attempt to build it, it should just work. If it fails, we've probably lost that potential contributor.

Pretty sure I can come up with an alternative check that doesn't disable cache for existing users though. I've since learned that the `aws` command we use does not seem to require authentication, despite mention of a private key in our docs, so we can instead check for an installed awscli and skip cache download if the `aws` command doesn't exist.

dberkholz replied:
Pushed a new commit as per the above.

nikita-b reviewed (changes requested):
Choose the wrong button :)

dberkholz committed:
Cache: flip default from false to autodetect.

Detect installation of awscli, and if it's there, use the S3 cache for
build-server-rpm. Still should work for non-Percona staff because the S3
cache does not require a private key.
